clang/test/CodeGenCoroutines/coro-await-suspend-destroy.cpp

```diff
@@ -0,0 +1,129 @@
+// RUN: %clang_cc1 -std=c++20 -triple x86_64-unknown-linux-gnu -emit-llvm -o - %s \
+// RUN:   -disable-llvm-passes | FileCheck %s --check-prefix=CHECK-INITIAL
+// RUN: %clang_cc1 -std=c++20 -triple x86_64-unknown-linux-gnu -emit-llvm -o - %s \
+// RUN:   -O2 | FileCheck %s --check-prefix=CHECK-OPTIMIZED
+
+#include "Inputs/coroutine.h"
+
+// Awaitable with `coro_await_suspend_destroy` attribute
+struct [[clang::coro_await_suspend_destroy]] DestroyingAwaitable {
+  bool await_ready() { return false; }
+  void await_suspend_destroy(auto& promise) {}
+  void await_suspend(auto handle) {
+    await_suspend_destroy(handle.promise());
+    handle.destroy();
+  }
+  void await_resume() {}
+};
+
+// Awaitable without `coro_await_suspend_destroy` (normal behavior)
+struct NormalAwaitable {
+  bool await_ready() { return false; }
+  void await_suspend(std::coroutine_handle<> h) {}
+  void await_resume() {}
+};
+
+// Coroutine type with `std::suspend_never` for initial/final suspend
+struct Task {
+  struct promise_type {
+    Task get_return_object() { return {}; }
+    std::suspend_never initial_suspend() { return {}; }
+    std::suspend_never final_suspend() noexcept { return {}; }
+    void return_void() {}
+    void unhandled_exception() {}
+  };
+};
+
+// Single co_await with coro_await_suspend_destroy.
+// Should result in no allocation after optimization.
+Task test_single_destroying_await() {
+  co_await DestroyingAwaitable{};
+}
+
+// CHECK-INITIAL-LABEL: define{{.*}} void @_Z28test_single_destroying_awaitv
+// CHECK-INITIAL: call{{.*}} @llvm.coro.alloc
+// CHECK-INITIAL: call{{.*}} @llvm.coro.begin
+
+// CHECK-OPTIMIZED-LABEL: define{{.*}} void @_Z28test_single_destroying_awaitv
+// CHECK-OPTIMIZED-NOT: call{{.*}} @llvm.coro.alloc
+// CHECK-OPTIMIZED-NOT: call{{.*}} malloc
+// CHECK-OPTIMIZED-NOT: call{{.*}} @_Znwm
```

Comment:
These checks are probably too simple. I would prefer to use matching get the variable name and do more pattern checking. 

A good example can be found here: https://github.com/llvm/llvm-project/blob/main/clang/test/CodeGenCoroutines/coro-await.cpp#L53

Reply:
@yuxuanchen1997, I'm not sure I understand your meaning. Can you make it more explicit?
 - What is bad about these checks being simple? Are you worried about insufficient coverage, or future failures that are due to a check not being specific enough?
     - If you want more coverage, can you suggest more tests?
     - If you're want me to make the tests more specific, how can I do that?
 - What variable name should I be testing for?

Reply:
Got it, thanks! So you're not specifically calling out these "no heap alloc" test lines with `CHECK-OPTIMIZED-NOT`. Rather, you want to see a skeleton of the coro in the IR, so the tests are a bit more self-documenting.

On the next rev, I'll give it a go, following the example you linked.

> read the IR

I've done that, of course, as part of writing and debugging these tests :)

On these particular lines, I just tried to write the shortest test that shows the desired behavior change of "no alloc".